src/lib/ares_parse_srv_reply.c

```diff
@@ -38,128 +39,76 @@ int
   if (status != ARES_SUCCESS)
+  {
+    if (csrv_out)
+      ares_free_data (csrv_out);
     return status;
-
-  if (aptr + len + QFIXEDSZ > abuf + alen)
+  }
+
+  /* iterate through the cares_srv_reply list and
+   * create a new ares_srv_reply */
+  for (csrv_curr = csrv_out; csrv_curr;
+    csrv_curr = cares_srv_reply_get_next(csrv_curr))
+  {
+    srv_curr = ares_malloc_data(ARES_DATATYPE_SRV_REPLY);
+    if (!srv_curr)
     {
-      ares_free (hostname);
-      return ARES_EBADRESP;
+      status = ARES_ENOMEM;
+      break;
     }
-  aptr += len + QFIXEDSZ;
-
-  /* Examine each answer resource record (RR) in turn. */
-  for (i = 0; i < ancount; i++)
+    if (srv_last)
+    {
+      srv_last->next = srv_curr;
+    }
+    else
     {
-      /* Decode the RR up to the data field. */
-      status = ares_expand_name (aptr, abuf, alen, &rr_name, &len);
-      if (status != ARES_SUCCESS)
-        {
-          break;
-        }
-      aptr += len;
-      if (aptr + RRFIXEDSZ > abuf + alen)
-        {
-          status = ARES_EBADRESP;
-          break;
-        }
-      rr_type = DNS_RR_TYPE (aptr);
-      rr_class = DNS_RR_CLASS (aptr);
-      rr_len = DNS_RR_LEN (aptr);
-      aptr += RRFIXEDSZ;
-      if (aptr + rr_len > abuf + alen)
-        {
-          status = ARES_EBADRESP;
-          break;
-        }
-
-      /* Check if we are really looking at a SRV record */
-      if (rr_class == C_IN && rr_type == T_SRV)
-        {
-          /* parse the SRV record itself */
-          if (rr_len < 6)
-            {
-              status = ARES_EBADRESP;
-              break;
-            }
-
-          /* Allocate storage for this SRV answer appending it to the list */
-          srv_curr = ares_malloc_data(ARES_DATATYPE_SRV_REPLY);
-          if (!srv_curr)
-            {
-              status = ARES_ENOMEM;
-              break;
-            }
-          if (srv_last)
-            {
-              srv_last->next = srv_curr;
-            }
-          else
-            {
-              srv_head = srv_curr;
-            }
-          srv_last = srv_curr;
-
-          vptr = aptr;
-          srv_curr->priority = DNS__16BIT(vptr);
-          vptr += sizeof(unsigned short);
-          srv_curr->weight = DNS__16BIT(vptr);
-          vptr += sizeof(unsigned short);
-          srv_curr->port = DNS__16BIT(vptr);
-          vptr += sizeof(unsigned short);
-
-          status = ares_expand_name (vptr, abuf, alen, &srv_curr->host, &len);
-          if (status != ARES_SUCCESS)
-            break;
-        }
-
-      /* Don't lose memory in the next iteration */
-      ares_free (rr_name);
-      rr_name = NULL;
-
-      /* Move on to the next record */
-      aptr += rr_len;
+      srv_head = srv_curr;
+    }
+    srv_last = srv_curr;
+
+    /* copy the host to newhost so we can free csrv_out */
+    char *newhost = ares_strdup(cares_srv_reply_get_host(csrv_curr));
+    if (!newhost) {
+      status = ARES_ENOMEM;
+      break;
     }
```

Comment:
1. Thanks for telling me about strdup, it is better and simplifies the logic. On the last commit, I was missing the "else" part to set the status to an error if the malloc failed, but it's simpler and clearer now.
2. There are 3 places where allocations occur: csrv_out at the top and then srv_curr and newhost inside the loop for each RR.  I think in each case I check for null, set an error, make sure all three of them are freed (newhost is freed as part of the new ares_srv_reply linked list), and return the error status. Does it look right to you?

Reply:
also, do you want me to add each RR as a separate PR?  Or put them all in one?